### ISSUE TITLE
Initialize the state synchronously with the values of the props

### ISSUE THREAD

**Issue by flaviendelangle** [mui-org/material-ui-x#2293]
> We could move the initial state to the hook level, instead of having a centralized place.

I don't have the whole pictures of the impacts this change would have.
But a few ideas :

## Approach 1

```ts
const useGridPageSize = (api: GridApiRef, props: Pick<GridComponentProps, 'pageSize, 'onPageSizeChange'>)  => {
  useGridStateInit(apiRef, state => ({ ...state, pagination: { page: props.page ?? 0 }}));
}

const useGridPage = (api: GridApiRef, props: Pick<GridComponentProps, 'page, 'onPageChange'>)  => {
  useGridStateInit(state => ({ ...state, pagination: { ...state.pagination!, pageSize: props.pageSize }}));
}
```

```ts
const useGridStateInit = (apiRef, callback: (state: Partial<GridState>): Partial<GridState>)) => {
  const isInitialized = React.useRef(false);
  const [, setGridState] = useGridState(apiRef);

  if (!isInitialized.current) {
    setGridState(callback);
  }
} 
```

We would have a synchronous state initialization with the value from the props and the hook being responsible for its own state.
But the typing `Partial<GridState>` is not very precise since its relying on the hook order.

The main drawback here is that  could not use selectors in the render that come from hooks later in the feature chain.
To avoid this we must have a state initializer in the hook file / folder, but not in the hook function so that it can be called before the hooks are called.
___

## Approach 2

Each hook can export a state initializer. All hooks are called in a single hook taking care of the state initialization and then running the hooks themselves..

```ts
// useGridPageSoze.ts
export const useGridPageSoze: GridPlugin<'pageSize', 'onPageSizeChange'> = (api, props) => {

}

useGridPage.initState = (state, props) => ({ ...state, pagination: { page: props.pageSize ?? 100 }})
```

```ts
// useGridPage.ts
export const useGridPage: GridPlugin<'page', 'onPageChange'> = (api, props) => {

}

useGridPage.initState = (state, props) => ({ ...state, pagination: { ...state.pagination!, page: props.page ?? 0 }})
```

```ts
// GridPlugin.ts
export type GridPlugin<PropKeys extends keyof GridComponentProps> = {
  (apiRef: GridApiRef, props: Pick<GridComponentProps, PropKeys>): void;
  initState?: (state: Partial<GridState>, props: Pick<GridComponentProps, PropKeys>) => Partial<GridState>
} 
```

```ts
// useGridPlugins.ts
// The plugin must change from a render to another to respect the rule of hooks
const useGridPlugins = (plugins: GridPlugin[]): void => {
  const isInitialized = React.useRef(false)

  if (!isInitialized.current) {
    isInitialized.current = true
    apiRef.current.state = plugins.reduce(
      (acc, plugin) => plugin?.initState(acc) ?? acc, 
      {} as Partial<GridState>
    ) 
    apiRef.current.publishEvent(GridEvents.initState, state)
  }

  for (const plugin in plugins) {
    plugin(apiRef, props);
  }
} 
```

```ts
const X_GRID_PLUGINS: GridPlugin[] = [useGridPageSize, useGridPage]

// XGrid.tsx
const XGrid = (inProps: XGridProps) => {
  const apiRef = useGridApiRef(inProps.apiRef);
  const props = useThemeProps({ props: inProps, name: 'MuiDataGrid' });

  useGridPlugins(X_GRID_PLUGINS)
}
```

With this approach, the 1st time the hooks are called, they already have an initialized state even though they are responsible for the initialization.

The `initState` could allow state mutation if we want since its only called once before any state usage.
___

## Approach 3

Each hook can export a state initializer. The `useXGridComponents` and `useDataGridComponents` are then responsible or initializing the state by manually calling each of those state initializer.
It avoids building a generic plugin system but is more error prone because the initialization order must be correct in both files.

```ts
// useGridPageSize.ts
export const useGridPage = (api: GridApiRef, props: Pick<GridComponentProps, 'page, 'onPageChange'>) => {
}

export const initPageSizeState = (state, props) => ({ ...state, pagination: { page: props.page ?? 0 }})
```

```ts
// useGridPage.ts
export const useGridPage = (api: GridApiRef, props: Pick<GridComponentProps, 'page, 'onPageChange'>) => {
}

export const initPageState = (state, props) => ({ ...state, pagination: { ...state.pagination!, page: props.page ?? 0 }})
```

```ts
// XGrid.tsx
const XGrid = (inProps: XGridProps) => {
  const apiRef = useGridApiRef(inProps.apiRef);
  const props = useThemeProps({ props: inProps, name: 'MuiDataGrid' });

  const isInitialized = React.useRef(false)

  if (!isInitialized.current) {
    isInitialized.current = true
   
    let state: Partial<GridState> = {};
    state = initPageSizeState(state, props);
    state = initiPageState(state, props);
    apiRef.current.state = state as GriState;
    apiRef.current.publishEvent(GridEvents.initState, state);
  }

  useGridPageSize(apiRef, props);
  useGridPage(apiRef, props);
}
```
___

This is a major change not to take lightly of course.

I think it tightly related to the prop-to-state pattern (https://github.com/mui-org/material-ui-x/issues/1562#issuecomment-860247688) and we may want to first take a deeper look at the approach we want to have for the update part.

In summary, maybe its better to wait to have a cleaner view of the approach we want to take.

**Comment by oliviertassinari** [comment]
> If we initialize the state synchronously, we will have to rework a bit the control state API. 

I would even say that it should fix:

https://github.com/mui-org/material-ui-x/blob/49639ee9924517b5f1aee3d8233f0ddd669c8fdc/packages/grid/_modules_/grid/hooks/features/core/useGridControlState.ts#L33-L35

Regarding the options. I would personally try 1 first, see if it's flying, otherwise fallback to 2. Option 3 seems to force to have to handle each plugin manually,

**Comment by flaviendelangle** [comment]
I agree on Option 3. I had to propose it since its a lighter version of Option 2 on the implementation but it's not viable for me.

For Option 1 to work, we need to be even stricter about the order of the hooks. 
I would be in favor of declaring explicitely in each hook which hooks it requires.

Maybe through the JSDoc

```ts
@requires usePageSize
const usePage = () => { ... } 
```

Or in TypeScript is we want to do a validation in the future for custom use hooks, but it may be overkill for now.

**Comment by flaviendelangle** [comment]
I'm doing a first working draft of the state synchronous initialization
Overall it seems to be working fine.


We just have a few points to address:

## Hook example

```ts
export const useGridPage = (
  apiRef: GridApiRef,
  props: Pick<GridComponentProps, 'page' | 'onPageChange' | 'rowCount'>,
) => {
  // 1. LOGGER
  const logger = useLogger('useGridPage');

  // 2. STATE INITIALIZATION
  // We initialize the state before 3. because any selector because if the hook is relying on selectors of its own state, it would crash
  useGridStateInit(apiRef, (state) => ({
    ...state,
    pagination: {
      ...state.pagination!,
      page: 0,
      pageCount: getPageCount(props.rowCount ?? 0, state.pagination!.pageSize!),
      rowCount: props.rowCount ?? 0,
    },
  }));

  // 3. STATE SELECTOR AND UPDATER
  const [, setGridState, forceUpdate] = useGridState(apiRef);
  const visibleRowCount = useGridSelector(apiRef, visibleGridRowCountSelector);

  // 4. CONTROL STATE REGISTRATION
  // We synchronously define the control state (that's one of the point that I still need to work on)
  useGridRegisterControlState(apiRef, {
    stateId: 'page',
    propModel: props.page,
    propOnChange: props.onPageChange,
    stateSelector: (state) => state.pagination.page,
    changeEvent: GridEvents.pageChange,
  });

  // 5. API METHODS
  const setPage = React.useCallback<GridPageApi['setPage']>(
    (page) => {
      logger.debug(`Setting page to ${page}`);

      setGridState((state) => ({
        ...state,
        pagination: applyValidPage({
          ...state.pagination,
          page,
        }),
      }));
      forceUpdate();
    },
    [setGridState, forceUpdate, logger],
  );

  const pageApi: GridPageApi = {
    setPage,
  };

  useGridApiMethod(apiRef, pageApi, 'GridPageApi');

  // 6. UPDATE EFFECTS
  // A major future evolution would be to limit as much as possible this pattern to avoid `useEffect` cascade.
  // TODO: Avoid having this `setGridState` run after the first render (either do manual test in the `useEffect` or create a `useEffectNotFirstRender` hook
  React.useEffect(() => {
    setGridState((state) => {
      const rowCount = props.rowCount !== undefined ? props.rowCount : visibleRowCount;
      const pageCount = getPageCount(rowCount, state.pagination.pageSize);

      const page = props.page == null ? state.pagination.page : props.page;

      return {
        ...state,
        pagination: applyValidPage({
          ...state.pagination,
          page,
          rowCount,
          pageCount,
        }),
      };
    });
    forceUpdate();
  }, [setGridState, forceUpdate, visibleRowCount, props.rowCount, props.page, apiRef]);

  // 7. EVENT CALLBACKS
  const handlePageSizeChange = React.useCallback(
    (pageSize: number) => {
      setGridState((state) => {
        const pageCount = getPageCount(state.pagination.rowCount, pageSize);

        return {
          ...state,
          pagination: applyValidPage({
            ...state.pagination,
            pageCount,
            page: state.pagination.page,
          }),
        };
      });

      forceUpdate();
    },
    [setGridState, forceUpdate],
  );

  useGridApiEventHandler(apiRef, GridEvents.pageSizeChange, handlePageSizeChange);
};
```

## Issues I faced

### Some components uses selectors of `useGridReorder` which is a pro-only hook.

Now that the hook is the one initialization its state, the free version has no state at all. Which make sense, as we move forward to a more plugable API, we can't have every version of the Grid contain a fake-initial state.

As @oliviertassinari pointed out, it feels wrong that the components needs to import selectors like this (see https://github.com/mui-org/material-ui-x/issues/924#issuecomment-891779404)
For now I bypassed the problem by creating a `useGridProSelector` which test the signature and return `undefined` if it is not the pro version rather than calling the selector. But it's not what we want in the final version.

### Some state initialization are too complicated to be done at the start of their hook

For instance, the filter state is applied like this

```ts
  React.useEffect(() => {
    // [...]
    const oldFilterModel = apiRef.current.state.filter;
    if (props.filterModel !== undefined && props.filterModel !== oldFilterModel) {
      logger.debug('filterModel prop changed, applying filters');
      setGridState((state) => ({
        ...state,
        filter: props.filterModel,
      }));
      apiRef.current.applyFilters();
    }
  }, [apiRef, logger, props.filterModel, setGridState]);
```

I added the `filter: props.filterModel` inside the initialization hook, but the `apiRef.current.applyFilters` can hardly be done at the start of the file because it needs like half of the hook code

So I did the following

```ts
export const useGridFilter = (apiRef, props): void => {
  const logger = useLogger('useGridFilter');

  useGridStateInit(apiRef, (state) => ({
    ...state,
    filter: props.filterModel ?? getInitialGridFilterState(),
    visibleRows: {
      visibleRowsLookup: {},
    },
  }));

  // ... all the rest of the hook code

    useFirstRender(() => apiRef.current.applyFilters());
}
```

Which means, if we have synchronous access to the state during the 1st render, `virtualRows` won't be initialized.
They may be a better solution and it could be an issue when trying synchronous state updates, but for now it's working fine.

**Comment by flaviendelangle** [comment]
@oliviertassinari @m4theushw  @DanailH when you have some time to take a look at the hook example above and the points to be improved.

https://github.com/flaviendelangle/material-ui-x/pull/1/files
I'll open this PR in the `mui-org` repo to start talk about the implementation and the impacts on specific feature hooks.
